src/v/kafka/server/group_manager.cc

```diff
@@ -567,6 +567,44 @@ group_manager::leave_group(leave_group_request&& r) {
     }
 }
 
+ss::future<txn_offset_commit_response>
+group_manager::txn_offset_commit(txn_offset_commit_request&& r) {
+    auto catchup_lock = get_catchup_lock(r.ntp);
```

Comment:
I'm still struggling with why catchup_lock is a shared pointer. for example, when is an attached partition in the _partition set with the associated catchup lock unset?

Would the following be equivalent:

```c++
struct attached_partition {
   ...
   // store as a value instead of shared pointer
   ss::basic_rwlock<> catchup_lock;
   ...
};

...

std::optional<ss::basic_rwlock<>> get_catchup_lock(ntp) {
   if (auto it = _partitions.find(ntp); it != _partitions.end()) {
     return it->second.catchup_lock;
   }
   return std::nullopt;
}

// that has the benefit of not requiring any changes to call sites of get_catchup_lock
// but the preferred interface would be more like:

lw_shared_ptr<attached_partition> get_partition(ntp) {
   if (auto it = _partitions.find(ntp); it != _partitions.end()) {
     return it->second;
   }
   return std::nullopt;
}

...

auto p = get_partition(r.ntp);
if (!p || !p->catchup_lock.try_read_lock()) {
   ...
}
```

of course if the catchup_lock actually can be nullptr inside an instance of attached_partition then this doesn't matter but I don't see any case where that is possible.

Reply:
good idea with `get_partition`, I've done it